### PR TITLE
chore: use cobra and viper for local-artifact-mirror

### PR DESCRIPTION
**Pull request by marccampbell**

### CODE REVIEW

**Review by sgalsaleh — APPROVED** [review]

File: cmd/local-artifact-mirror/pull_binaries.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_binaries.go:36]: this isnt needed

File: cmd/local-artifact-mirror/pull_helmcharts.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_helmcharts.go:35]: this isnt needed

File: cmd/local-artifact-mirror/pull_images.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_images.go:35]: this isnt needed

File: cmd/local-artifact-mirror/serve.go [hunk 4]
```diff
@@ -25,81 +25,96 @@ var (
...
-		var provider *defaults.Provider
-		if c.IsSet("data-dir") {
-			provider = defaults.NewProvider(c.String("data-dir"))
-		} else {
-			var err error
-			provider, err = cmdutil.NewProviderFromFilesystem()
-			if err != nil {
-				panic(fmt.Errorf("unable to get provider from filesystem: %w", err))
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:47]: this isnt needed

File: cmd/local-artifact-mirror/pull.go [hunk 5]
```diff
@@ -32,189 +27,31 @@ const (
...
+			}
+			kubecli = kc
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.Help()
+			os.Exit(1)
+			return nil
+		},
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull.go:47]: i feel like you can omit this entirely but maybe you tried that

File: cmd/local-artifact-mirror/serve.go [hunk 6]
```diff
@@ -25,81 +25,96 @@ var (
...
+		dataDir string
+		port    int
+	)
+
+	cmd := &cobra.Command{
+		Use:   "serve",
+		Short: "Serve files from the data directory over HTTP",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+			v.BindPFlag("port", cmd.Flags().Lookup("port"))
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:39]: you could shorten this to `v.BindPFlags(cmd.Flags())`

File: cmd/local-artifact-mirror/pull_binaries.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// binariesCommands pulls the binary artifact from the registry running in the cluster and stores
+// it locally. This command is used during cluster upgrades when we want to fetch the most up to
+// date binaries. The binaries are stored in the /usr/local/bin directory and they overwrite the
+// existing binaries.
+func PullBinariesCmd(ctx context.Context, v *viper.Viper) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "binaries <installation-name>",
+		Short: "Pull binaries artifacts for an airgap installation",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_binaries.go:28]: you could shorten this to `v.BindPFlags(cmd.Flags())`

File: cmd/local-artifact-mirror/pull_helmcharts.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+// helmChartsCommand pulls helm charts from the registry running in the cluster and
+// stores them locally. This command is used during cluster upgrades when we want to
+// fetch the most up to date helm charts. Helm charts are stored in a tarball file
+// in the default location.
+func PullHelmChartsCmd(ctx context.Context, v *viper.Viper) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "helmcharts <installation-name>",
+		Short: "Pull Helm chart artifacts for an airgap installation",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_helmcharts.go:27]: you could shorten this to `v.BindPFlags(cmd.Flags())`

File: cmd/local-artifact-mirror/pull_helmcharts.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_helmcharts.go:31]: Could you use this instead? https://pkg.go.dev/github.com/spf13/cobra#ExactArgs

File: cmd/local-artifact-mirror/pull_images.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+func PullImagesCmd(ctx context.Context, v *viper.Viper) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "images <installation-name>",
+		Short: "Pull images artifacts for an airgap installation",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_images.go:31]: Could you use this instead? https://pkg.go.dev/github.com/spf13/cobra#ExactArgs

File: cmd/local-artifact-mirror/pull_binaries.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+			if len(args) != 1 {
+				return errors.New("expected installation name as argument")
+			}
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_binaries.go:32]: Could you use this instead? https://pkg.go.dev/github.com/spf13/cobra#ExactArgs

File: cmd/local-artifact-mirror/pull_images.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+// imagesCommand pulls images from the registry running in the cluster and stores
+// them locally. This command is used during cluster upgrades when we want to fetch
+// the most up to date images. Images are stored in a tarball file in the default
+// location.
+func PullImagesCmd(ctx context.Context, v *viper.Viper) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "images <installation-name>",
+		Short: "Pull images artifacts for an airgap installation",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_images.go:27]: you could shorten this to `v.BindPFlags(cmd.Flags())`

File: cmd/local-artifact-mirror/main.go [hunk 13]
```diff
@@ -6,26 +6,57 @@ import (
...
+func InitAndExecute(ctx context.Context, name string) {
+	v := viper.GetViper()
+	if err := RootCmd(ctx, v, name).Execute(); err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 }
+
+func initConfig(v *viper.Viper) {
+	v.SetEnvPrefix("REPLICATED")
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/main.go:60]: You are going to have to change these to REPLICATED_DATA_DIR and REPLICATED_PORT.

The port one feels a little overly generic but your call.

https://github.com/replicatedhq/embedded-cluster/blob/fbc6db8672e10f18d811e50d9be1df8c162ab2e8/pkg/cmd/install.go#L90-L91

File: cmd/local-artifact-mirror/pull_binaries.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+			dataDir := v.GetString("data-dir")
+
+			// Support for older env vars
+			flag := cmd.Flags().Lookup("data-dir")
+			if flag == nil || !flag.Changed {
+				if os.Getenv("LOCAL_ARTIFACT_MIRROR_DATA_DIR") != "" {
+					dataDir = os.Getenv("LOCAL_ARTIFACT_MIRROR_DATA_DIR")
+				}
+			}
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/pull_binaries.go:46]: could you make a convenience function for this since its repeated 3 times?

File: cmd/local-artifact-mirror/serve.go [hunk 15]
```diff
@@ -25,81 +25,96 @@ var (
...
-		} else {
-			var err error
-			provider, err = cmdutil.NewProviderFromFilesystem()
-			if err != nil {
-				panic(fmt.Errorf("unable to get provider from filesystem: %w", err))
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
+
+			// there are some env vars that we still support for backwards compatibility
+			if os.Getenv("LOCAL_ARTIFACT_MIRROR_PORT") != "" {
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:50]: Have you tried v.BindEnv for this?

https://pkg.go.dev/github.com/spf13/viper#BindEnv

File: cmd/local-artifact-mirror/serve.go [hunk 16]
```diff
@@ -25,81 +25,96 @@ var (
...
+				envvarPort, err := strconv.ParseInt(os.Getenv("LOCAL_ARTIFACT_MIRROR_PORT"), 10, 32)
+				if err != nil {
+					return fmt.Errorf("unable to parse LOCAL_ARTIFACT_MIRROR_PORT: %w", err)
+				}
+				port = int(envvarPort)
+			}
+			if os.Getenv("LOCAL_ARTIFACT_MIRROR_DATA_DIR") != "" {
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:57]: Have you tried v.BindEnv for this?

https://pkg.go.dev/github.com/spf13/viper#BindEnv

File: cmd/local-artifact-mirror/serve.go [hunk 17]
```diff
@@ -25,81 +25,96 @@ var (
...
 			}
-		}
 
-		port, err := k8snet.ParsePort(c.String("port"), false)
-		if err != nil {
-			panic(fmt.Errorf("unable to parse port from flag: %w", err))
-		}
+			var provider *defaults.Provider
+			if v.Get("data-dir") != nil {
+				provider = defaults.NewProvider(v.GetString("data-dir"))
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:63]: This does not respect the env var `LOCAL_ARTIFACT_MIRROR_DATA_DIR`. You need to use the `dataDir` variable

File: cmd/local-artifact-mirror/serve.go [hunk 18]
```diff
@@ -25,81 +25,96 @@ var (
...
+				var err error
+				provider, err = cmdutil.NewProviderFromFilesystem()
+				if err != nil {
+					panic(fmt.Errorf("unable to get provider from filesystem: %w", err))
+				}
+			}
 
-		os.Setenv("TMPDIR", provider.EmbeddedClusterTmpSubDir())
+			port := v.GetInt("port")
+			if port == 0 {
```
Comment (emosbaugh) [inline cmd/local-artifact-mirror/serve.go:73]: this overwrites the env var

File: cmd/local-artifact-mirror/serve.go (hunk 17 above)
Comment (emosbaugh) [reply cmd/local-artifact-mirror/serve.go:63]: i think we could really benefit from moving the flag and env var parsing to functions and write unit tests

File: cmd/local-artifact-mirror/serve.go (hunk 18 above)
Comment (emosbaugh) [reply cmd/local-artifact-mirror/serve.go:73]: i think we could really benefit from moving the flag and env var parsing to functions and write unit tests

File: cmd/local-artifact-mirror/serve.go (hunk 16 above)
Comment (emosbaugh) [reply cmd/local-artifact-mirror/serve.go:57]: This will not work

File: cmd/local-artifact-mirror/serve.go (hunk 15 above)
Comment (emosbaugh) [reply cmd/local-artifact-mirror/serve.go:50]: This will not work
